sway-fmt-v2/src/utils/comments.rs

```diff
@@ -69,6 +100,308 @@ fn collect_comments_from_token_stream(
+            previous_unformatted_comment_span,
+            unformatted_comment_span,
+            &comment_map,
+            &unformatted_code,
+        );
+        if !comments_found.is_empty() {
+            offset += insert_after_span(
+                previous_formatted_comment_span,
+                comments_found,
+                offset,
+                formatted_code,
+            )?;
+        }
+        previous_unformatted_comment_span = unformatted_comment_span;
+        previous_formatted_comment_span = formatted_comment_span;
+    }
+    Ok(())
+}
+
+// A CommentWithContext is the Comment and the offset before it. The offset can be between the (from) item we searched for this comment or from the last comment inside range
+type CommentWithContext = (Comment, String);
+
+/// Returns a list of comments between given spans. For each comment returns the Context
+/// Context of a comment is basically the offset (the characters between the last item/comment) to the current comment
+fn get_comments_between_spans(
+    from: &CommentSpan,
+    to: &CommentSpan,
+    comment_map: &CommentMap,
+    unformatted_code: &Arc<str>,
+) -> Vec<CommentWithContext> {
+    let mut comments_with_context = Vec::new();
+    for (index, comment_tuple) in comment_map
+        .range((Included(from), Excluded(to)))
+        .enumerate()
+    {
+        if comments_with_context.is_empty() {
+            // This is the first comment in the current range the context should be collected between from's end and comment's beginning
+            comments_with_context.push((
+                comment_tuple.1.clone(),
+                unformatted_code[from.end..comment_tuple.0.start].to_string(),
+            ));
+            println!("this");
+        } else {
+            // There is a comment before this one, so we should get the context starting from the last comment's end to the beginning of the current comment
+            comments_with_context.push((
+                comment_tuple.1.clone(),
+                unformatted_code
+                    [comments_with_context[index - 1].0.span.end()..comment_tuple.0.start]
+                    .to_string(),
+            ));
+        }
+    }
+    println!(
+        " found {:?} between {:?} and {:?} \n\n",
+        comments_with_context, from, to
+    );
+    comments_with_context
+}
+
+/// Inserts after given span and returns the offset. While inserting comments this also inserts Context of the comments so that the alignment whitespaces/newlines are intact
+fn insert_after_span(
+    from: &CommentSpan,
+    comments_to_insert: Vec<CommentWithContext>,
+    offset: usize,
+    formatted_code: &mut FormattedCode,
+) -> Result<usize, FormatterError> {
+    let mut src_rope = Rope::from_str(formatted_code);
+    // prepare the comment str
+    let mut comment_str = format!(
+        "{}{}",
+        comments_to_insert[0].1,
+        format_comment(&comments_to_insert[0].0)
```

Comment:
This `insert_after_span` function could be made a little safer if we were to use `if let` to remove the assumption that the `comments_to_insert` is not empty, e.g.

```rust
let mut iter = comments_to_insert.iter();
if let Some(first) = iter.next() {
    let mut comment_str = format!("{}{}", first.1, format_comment(&first.0));
    for comment in iter {
        // ...
    }
    return Ok(comment_str.len());
}
Ok(0)